### DISCUSSION TITLE
Help with freezing d1mini & 7 segment displays

### DISCUSSION THREAD

**Discussion by mrbc42** [micropython/micropython#10536] [General]
Hi all, 
I have a D1-mini with MP 1.19.1installed.  It has two 7 segment displays attached working as a clock but each in a different timezone. Essentially there is nothing that is not working in my code, the issue I am having is it will work for a few hours and then freeze.  Was hoping someone would be able to point something out as to why the freeze / crash is happening and how to resolve it.
When the problem occurs, both clocks just stay on at the time and don't increment.  A reboot solves it till the next time.
thanks for any suggestions

Using tm1637.py for the display driver  - https://github.com/mcauser/micropython-tm1637
```
# Boot.py
import esp
esp.osdebug(None)
#import uos, machine, network
import machine, network

ssid = 'xxxx'
password = 'xxxx'

station = network.WLAN(network.STA_IF)

station.active(True)
station.connect(ssid, password)
```

```
# main.py
import tm1637
from machine import Pin
nsw = tm1637.TM1637(clk=Pin(12), dio=Pin(14))
qld = tm1637.TM1637(clk=Pin(04), dio=Pin(00))
import ntptime
import time
import utime
from umqtt.robust import MQTTClient

ntptime.settime()
nsw.brightness(0)
qld.brightness(0)

ntptime.host = "time.google.com"
x = 0

client_id = 'clock_test'
mqtt_server = '192.168.5.45'

def mqtt_connect():
    client = MQTTClient("client_id", mqtt_server, port=1883, keepalive=10)
    client.connect()
    return client

while True:
    client = mqtt_connect()
    nswclock_hour = utime.localtime(utime.mktime(utime.localtime()) + 11*3600)[3]
    nswclock_min = utime.localtime(utime.mktime(utime.localtime()) + 11*3600)[4]
    
    qldclock_hour = utime.localtime(utime.mktime(utime.localtime()) + 10*3600)[3]
    qldclock_min = utime.localtime(utime.mktime(utime.localtime()) + 10*3600)[4]
 
    if nswclock_min == 01 and x == 0:
        ntptime.settime()
        x += 1
    if nswclock_min == 02:
        x = 0
        
    nsw.numbers(nswclock_hour, nswclock_min)
    qld.numbers(qldclock_hour, qldclock_min)
    client.publish('clock_digits', str(nswclock_min))
    client.disconnect()
    time.sleep(30) 
    
```

**Comment by peterhinch** [comment]
I think it's likely that you're getting brief WiFi outages. See [this guide](https://github.com/peterhinch/micropython-samples/blob/master/resilient/README.md). Also [this resilient MQTT library](https://github.com/peterhinch/micropython-mqtt).

**Reply by mrbc42** [reply]
Thanks, I will try this.

**Comment by mrbc42** [comment]
Hi Peter, I disabled all MQTT code and it ran fine for over 24 hours so your diagnosis was correct, thank you.  I downloaded your library and compiled it to mpy and put it on the board but I am stuck with the config.  Although your documentation is thorough, I cannot see a simple example where I can publish to my server.

Could you provide a small example or point to something I have missed?  Using this code, I get an error.  I can see in the docs there are many more config options so unsure why i am getting an error about 3 arguments.
thanks in advance.

```
from mqtt_as import MQTTClient
import uasyncio as asyncio

client_id = 'clock_test'
mqtt_server = '192.168.5.45'

def mqtt_connect():
    client = MQTTClient("client_id", mqtt_server, port=1883, keepalive=10)
    client.connect()
    return client

while True:
    client = mqtt_connect()

```
TypeError: function takes 3 positional arguments but 4 were given

**Reply by peterhinch** [reply]
This has a number of problems. The client constructor takes a single arg which is a `dict`. You set parameters by changing the `dict` entries before instantiating the client. But above all the library is **asynchronous**: applications must be written as asynchronous code using `uasyncio`. Please study the examples such as [range.py](https://github.com/peterhinch/micropython-mqtt/blob/master/mqtt_as/range.py).

As a new user I'd advise progressively adapting one of the existing examples rather than trying to start from scratch with new code.
